### DISCUSSION TITLE
Can Whisper be installed using Poetry?

### DISCUSSION THREAD

**Discussion by mywristbands** [openai/whisper#683] [Q&A]
I'm using Poetry to manage my python package dependencies, and I'd like to install Whisper. I tried doing this by adding the following line to my `pyproject.toml` file:
```
whisper = {git = "https://github.com/openai/whisper.git", rev = "0b5dcfdef7ec04250b76e13f1630e32b0935ce76"}
```

And then I ran `poetry update whisper`.

While this does work to install Whisper, it doesn't install any of Whisper's dependencies listed in `requirements.txt`. I could manually download all of these dependencies, but this is really something my package manager is supposed to do. Is it possible to use Poetry to automatically download whisper _and_ its dependencies?

If Poetry can't do the job here, which other python dependency manager would work? My preferred alternative is Pipenv, or next would be Conda.

**Comment by Vuizur** [comment]
What seems to be happening here is that poetry does not work with the `setup.py` of the whisper repo. Usually the `setup.py` is written independently of `requirements.txt` and has less strict version restrictions. However, whisper uses some technique to automatically generate the `setup.py`'s `install_requires` option out of the requirements.txt. It might be a good idea to open an issue in the poetry repository.

There is still a workaround that works if you're using poetry. You could execute `poetry shell` and then `pip install git+<linktorepo>`. Or you could install all dependencies listed in the requirements.txt using `poetry add <dependency>` manually.

**Reply by mywristbands** [reply]
Thanks @Vuizur for providing that context. I ended up (almost) getting it to work for now by using this [unofficial distribution of whisper](https://pypi.org/project/whisper.ai/). Here's what I've done:

I added the following to my myproject.toml file:
```
torch = [
  {markers = "sys_platform == 'darwin'", url = "https://download.pytorch.org/whl/cpu/torch-1.13.0-cp310-none-macosx_11_0_arm64.whl"},
  {markers = "sys_platform == 'linux' and platform_machine == 'aarch64'", url="https://download.pytorch.org/whl/torch-1.13.0-cp310-cp310-manylinux2014_aarch64.whl"},
  {markers = "sys_platform == 'linux' and platform_machine == 'x86_64'", url="https://download.pytorch.org/whl/cu116/torch-1.13.0%2Bcu116-cp310-cp310-linux_x86_64.whl"}
]
```
And then I ran `poetry update torch`. Then I ran `poetry add whipser.ai`, which installed whisper and all its remaining dependencies. I had to manually install torch first because whenever I ran `poetry add whisper.ai`, it would always throw an error related to torch.

Note that while this is working locally on my Mac with the M1 Pro chip, as well as in my Linux docker container running locally on my Mac, I've yet to get it running in our AWS GPU server. I have a support case open with AWS to hopefully resolve this.

**Comment by mywristbands** [comment]
For anyone reading this thread, ultimately I gave up on using Poetry, and instead decided to manage my python package dependencies with pip, pyenv, pyenv-virtualenv, and a `requirements.txt` file. Here's how I set it up (on macOS):

1. In a terminal window, cd into your project folder.
2. Install pyenv by following steps 1 and 2 under [Homebrew in macOS](https://github.com/pyenv/pyenv#homebrew-in-macos). Note that for step 2, make sure you're setting up the shell environment for whatever shell(s) you actually use.
3. Run `brew install pyenv-virtualenv` // Installs the pyenv-virtualenv plugin for more easily managing python environments.
4. Run `echo 'eval "$(pyenv virtualenv-init -)"' >> ~/.zshrc`. // Adds pyenv-virtualenv to your `$PATH` for easier command-line usage. Note: If you use bash for your terminal instead of zsh, use `~/.profile` instead of `~/.zshrc` in the command.
5. Run `pyenv install <your.python.version>` // Installs the python version you'd like to use for your project. Replace `<your.python.version>` with the actual python version you'd like to use, such as `3.10.9`.
6. Run `pyenv virtualenv <your.python.version> <environment name>` // Creates virtual environment with desired python version and name of your choosing.
7. Run `pyenv local <environment name>` // Indicates to pyenv that this project uses the virtual environment we just created.
8. Restart the terminal. The specific command to do this depends on which shell you're using (see [here](https://stackoverflow.com/questions/57262349/restart-terminal-without-closing-on-macos)).
9. Now, whenever you cd into your project folder, it'll automatically activate the virtual environment!
10. From here (within your project folder), you can install all your desired python libraries by running `pip install <library name>`.
11. To record these installed dependencies into a `requirements.txt` (such as to allow project collaborators to easily install these dependencies into their own environment), run `pip freeze > requirements.txt`. Then, your collaborators can install all the listed project dependencies by running `pip install -r requirements.txt`.

Note that since whisper is installed from a git repository (not from PyPi), you'll need to edit the `requirements.txt` file to indicate the repository and desired commit to install whisper from. For example, if you want to install commit `0b5dcfd` of the whisper repository, you'd replace `whisper==1.0` in your `requirements.txt` file with `git+https://github.com/openai/whisper.git@0b5dcfd#egg=whisper`.

Hope this is helpful for anyone who couldn't get whisper installed with Poetry and is looking for a decent alternative.

**Reply by dimitryzub** [reply]
Thank you @mywristbands for writing this 🙂

As an additional comment, an explicit example of how to change `whisper==1.0` in `requirements.txt` to install from Git repo to someone who didn't do it before:

```
# change this 
whisper==1.0

# to this
git+https://github.com/openai/whisper.git@main#egg=whisper
```

So it becomes:

```
reqeusts==2.3.5
bs4==1.0.0
git+https://github.com/openai/whisper.git@main#egg=whisper
...
```

- `@`: symbol defines from where to pull data, in this case from the `main` branch.
- `#`: hashtag symbol followed by `egg=whisper` is an explicitly written name of the package.

Reference: https://stackoverflow.com/a/35998253/15164646

**Comment by captnswing** [comment]
I managed to install whisper using poetry by adding 

```
openai-whisper = { git = "https://github.com/openai/whisper.git", rev = "v20231117" }
```

to my `pyproject.toml` file

**Reply by boyko11** [reply]
You just ended 40 minutes of misery for me. Thank you!
